ntoskrnl/io/iomgr/deviface.c

```diff
@@ -30,185 +31,421 @@ static PWCHAR BaseKeyString = L"\\Registry\\Machine\\System\\CurrentControlSet\\
-    PWCHAR StartPosition, EndPosition;
-    HANDLE ClassesKey;
-    PHANDLE GuidKeyRealP, DeviceKeyRealP, InstanceKeyRealP;
-    HANDLE GuidKeyReal, DeviceKeyReal, InstanceKeyReal;
+    UNICODE_STRING SymbolicLink;
     NTSTATUS Status;
 
-    SubKeyName.Buffer = NULL;
+    /* Build up new symolic link */
+    SymbolicLink.Length = 0;
+    SymbolicLink.MaximumLength = PrefixString->Length + DeviceString->Length + sizeof(L"#") + GuidString->Length + ReferenceString->Length + sizeof(UNICODE_NULL);
+    SymbolicLink.Buffer = ExAllocatePoolWithTag(NonPagedPool, SymbolicLink.MaximumLength, TAG_IO);
+    if (!SymbolicLink.Buffer)
+    {
+        DPRINT1("ExAllocatePoolWithTag() failed\n");
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
 
-    if (GuidKey != NULL)
-        GuidKeyRealP = GuidKey;
-    else
-        GuidKeyRealP = &GuidKeyReal;
+    Status = RtlAppendUnicodeStringToString(&SymbolicLink, PrefixString);
+    NT_ASSERT(NT_SUCCESS(Status));
 
-    if (DeviceKey != NULL)
-        DeviceKeyRealP = DeviceKey;
+    Status = RtlAppendUnicodeStringToString(&SymbolicLink, DeviceString);
+    NT_ASSERT(NT_SUCCESS(Status));
+
+    Status = RtlAppendUnicodeToString(&SymbolicLink, L"#");
+    NT_ASSERT(NT_SUCCESS(Status));
+
+    Status = RtlAppendUnicodeStringToString(&SymbolicLink, GuidString);
+    NT_ASSERT(NT_SUCCESS(Status));
+
+    if (ReferenceString != NULL)
+    {
+        Status = RtlAppendUnicodeStringToString(&SymbolicLink, ReferenceString);
+        NT_ASSERT(NT_SUCCESS(Status));
+    }
     else
-        DeviceKeyRealP = &DeviceKeyReal;
+    {
+        Status = RtlAppendUnicodeToString(&SymbolicLink, L"\\");
+        NT_ASSERT(NT_SUCCESS(Status));
+    }
+
+    DPRINT("New symbolic link is '%wZ'\n", &SymbolicLink);
+
+    *SymbolicLinkName = SymbolicLink;
+    return STATUS_SUCCESS;
+}
+
+static
+NTSTATUS
+IopSeparateSymbolicLink(_In_ PCUNICODE_STRING SymbolicLinkName,
+                        _Out_opt_ PUNICODE_STRING PrefixString,
+                        _Out_opt_ PUNICODE_STRING MungedString,
+                        _Out_opt_ PUNICODE_STRING GuidString,
+                        _Out_opt_ PUNICODE_STRING ReferenceString,
+                        _Out_opt_ PBOOLEAN ReferenceStringPresent,
+                        _Out_opt_ LPGUID InterfaceClassGuid)
+{
+    UNICODE_STRING KernelModePrefix = RTL_CONSTANT_STRING(L"\\??\\");
+    UNICODE_STRING UserModePrefix = RTL_CONSTANT_STRING(L"\\\\?\\");
+    UNICODE_STRING MungedPrefix = RTL_CONSTANT_STRING(L"##?#");
+    UNICODE_STRING MungedStringReal, GuidStringReal, ReferenceStringReal;
+    UNICODE_STRING DeviceString;
+    UNICODE_STRING LinkNameNoPrefix;
+    USHORT i, ReferenceStringOffset;
+    NTSTATUS Status;
+
+    MungedStringReal.Buffer = GuidStringReal.Buffer = ReferenceStringReal.Buffer = NULL;
+
+    DPRINT("Symbolic link is '%wZ'\n", SymbolicLinkName);
+
+    /* Symbolic link name looks like \??\ACPI#PNP0501#1#{GUID}\ReferenceString */
```

Comment:
It might be useful to use this example in comments throughout and explain that e.g. `DeviceString` would be `ACPI#PNP0501#1#` and so on

Reply:
Improved other comments in this function.